sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileAsyncClient.java

```diff
@@ -148,6 +183,368 @@ public String getFileName() {
+            DataLakeRequestConditions validatedRequestConditions = requestConditions == null
+                ? new DataLakeRequestConditions() : requestConditions;
+            /* Since we are creating a file with the request conditions, everything but lease id becomes invalid
+             after creation, so remove them for the append/flush calls. */
+            DataLakeRequestConditions validatedUploadRequestConditions = new DataLakeRequestConditions()
+                .setLeaseId(validatedRequestConditions.getLeaseId());
+            final ParallelTransferOptions validatedParallelTransferOptions =
+                ModelHelper.populateAndApplyDefaults(parallelTransferOptions);
+            long fileOffset = 0;
+
+            Function<Flux<ByteBuffer>, Mono<Response<PathInfo>>> uploadInChunksFunction = (stream) ->
+                uploadInChunks(stream, fileOffset, length, validatedParallelTransferOptions, headers,
+                    validatedUploadRequestConditions);
+
+            BiFunction<Flux<ByteBuffer>, Long, Mono<Response<PathInfo>>> uploadFullMethod =
+                (stream, lengthUploaded) -> uploadWithResponse(ProgressReporter
+                        .addProgressReporting(stream, validatedParallelTransferOptions.getProgressReceiver()),
+                    fileOffset, length, headers, validatedUploadRequestConditions);
+
+            return createWithResponse(null, null, headers, metadata, requestConditions)
+                .then(UploadUtils.determineUploadFullOrChunked(data, validatedParallelTransferOptions,
+                uploadInChunksFunction, uploadFullMethod));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    private Mono<Response<PathInfo>> uploadInChunks(Flux<ByteBuffer> data, long fileOffset, long length,
+        ParallelTransferOptions parallelTransferOptions, PathHttpHeaders httpHeaders,
+        DataLakeRequestConditions requestConditions) {
+        // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
+        AtomicLong totalProgress = new AtomicLong();
+        Lock progressLock = new ReentrantLock();
+
+        // Validation done in the constructor.
+        UploadBufferPool pool = new UploadBufferPool(parallelTransferOptions.getNumBuffers(),
+            parallelTransferOptions.getBlockSize(), MAX_APPEND_FILE_BYTES);
+
+        Flux<ByteBuffer> chunkedSource = UploadUtils.chunkSource(data, parallelTransferOptions);
+
+        /*
+         Write to the pool and upload the output.
+         */
+        return chunkedSource.concatMap(pool::write)
+            .concatWith(Flux.defer(pool::flush))
+            /* Map the data to a tuple, writing in the buffer.remaining temporarily */
+            .map(buffer -> Tuples.of(buffer, (long) buffer.remaining()))
+            /* The tuple keeps track of the next buffer to write and the fileOffset for the next buffer */
+            .scan((result, source) -> {
```

Comment:
Can you give a more thorough description of what this scan method is doing/how it works? Yes, there's the reactor docs, but perhaps walking through a few iterations in our own example will help improve clarity.

Reply:
made a Tuple 3, and added a block comment.